### DISCUSSION TITLE
[Proposal]: Spread operator for object initialization/assignment

### DISCUSSION THREAD

**Discussion by bagoette** [dotnet/csharplang#7507] [Language Ideas]
JavaScript has a [Spread Operator Syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) that, in my opinion, offers a really nice developer experience for object manipulation.

## Proposal

I would like to propose an object spread operator in `C#` for both object initialization and assignment
- Class examples
```csharp
public class Test1
{
    public int A { get; set; }
    public int B { get; set; }
}

public class Test2
{
    public int A { get; set; }
    public int B { get; set; }
    public int C { get; set; }
    public int D { get; set; }
}
```

1. Object Initialization (`source` and `destination` types are the same)
```csharp
var obj1 = new Test1 { A = 1, B = 2, };

var obj2 = ..obj1; // Creates new instance of Test1
// Or
var obj2 = { ..obj1 };
// Or
var obj2 = new Test1 { ..obj1 }; // Maybe explict `new Test1` for clarity?

// Lowered to
Test1 obj2 = new Test1 
{
    A = obj1.A,
    B = obj1.B,
};
```
2. Object Initialization (`source` and `destination` types are different)
  - As seen in the examples below, I feel like this feature can be type agnostic. By that I mean the `source` and `destination` types themselves shouldn't matter, only the property's types and names should be considered (and access modifiers)
  - When spreading an object of one type into another, the lowered code would just assign the matching properties (type and name)
  - The compiler could show errors or warnings when: 
    - There are no matching properties
    - The `destination` has `required` properties that the source does not
    - The `destination` has matching properties but are `getter` only
```csharp
var obj1 = new Test1 { A = 1, B = 2, };

Test2 obj2 = ..obj1;
// Or
Test2 obj2 = { ..obj1 };

// Lowered to
Test2 obj2 = new Test2
{
    A = obj1.A,
    B = obj1.B,
};

// Reversing types should still work
var obj1 = new Test2 { A = 1, B = 2, C = 3, D = 4 };

Test1 obj2 = ..obj1;
// Or
Test1 obj2 = { ..obj1 };

// Lowered to
Test1 obj2 = new Test1
{
    A = obj1.A,
    B = obj1.B,
};

// Required property example
public class TestRequired
{
    public int A { get; set; }
    public int B { get; set; }
    required public int C { get; set; }
    required public int D { get; set; }
}

var obj1 = new Test1 { A = 1, B = 2, };

TestRequired obj2 = ..obj1; // Compiler error

TestRequired obj2 = { // Valid
    ..obj1,
    C = 3,
    D = 4,
};

// Lowered to
TestRequired obj2 = new TestRequired
{
    A = obj1.A,
    B = obj1.B,
    C = 3,
    D = 4,
}
```
3. Object re-assignment
  - Similar to `JavaScript`, when you want to mutate an existing object, you should be able to spread the `destination` as part of the assignment to let the compiler know you're overwriting from the `source`
  -  I don't think this would need to create a new object, but not sure if this would cause syntax inconsistencies
```csharp
public class TestB
{
    public int B { get; set; }
}

var obj1 = new Test1 { A = 1, B = 2 };
var obj2 = new TestB { B = 6 };

obj1 = {
    ..obj1, // Keeps current obj1's properties intact
    ..obj2, // Overwrites B
};

// Lowered to
obj1.B = obj2.B;
```

I think this could reduce the need for reflection based auto mapping libraries and also reduce refactoring changes when a class has properties added or removed.

**Comment by TahirAhmadov** [comment]
This sounds like a request to add property mapping to C#, which has already been requested a good number of times and rejected, because property mapping is a job for a library method, not the language.
https://github.com/dotnet/csharplang/discussions/7393
https://github.com/dotnet/csharplang/discussions/7448
https://github.com/dotnet/csharplang/discussions/6778

Related:
https://github.com/dotnet/csharplang/discussions/5798
https://github.com/dotnet/csharplang/discussions/830
